svcb.go

```diff
@@ -0,0 +1,737 @@
+	l, _ := c.Next()
+	i, e := strconv.ParseUint(l.token, 10, 16)
+	if e != nil || l.err {
+		return &ParseError{"", "bad svb Priority", l}
+	}
+	rr.Priority = uint16(i)
+
+	c.Next()        // zBlank
+	l, _ = c.Next() // zString
+	rr.Target = l.token
+
+	name, nameOk := toAbsoluteName(l.token, o)
+	if l.err || !nameOk {
+		return &ParseError{"", "bad svcb Target", l}
+	}
+	rr.Target = name
+
+	// Values (if any)
+	l, _ = c.Next()
+	var xs []SVCBKeyValue
+	// Helps require whitespace between pairs.
+	// Prevents key1000="a"key1001=...
+	canHaveNextKey := true
+	for l.value != zNewline && l.value != zEOF {
+		switch l.value {
+		case zString:
+			if !canHaveNextKey {
+				// The key we can now read was probably meant to be
+				// a part of the last value.
+				return &ParseError{"", "svcb invalid value quotation", l}
+			}
+
+			// In key=value pairs, value doesn't have to be quoted
+			// unless value contains whitespace.
+			// And keys don't need to include values.
+			// Similarly, keys with an equality signs
+			// after them don't need values.
+			z := l.token
+
+			// z includes at least up to the first equality sign
+			idx := strings.IndexByte(z, '=')
+			key := ""
+			val := ""
+			var key_value SVCBKeyValue
+			if idx == -1 {
+				// Key with no value and no equality sign
+				key = z
+			} else if idx == 0 {
+				return &ParseError{"", "no valid svcb key found", l}
+			} else {
+				key = z[0:idx]
+				val = z[idx+1:]
+
+				if len(val) == 0 {
+					// We have a key and an equality sign
+					// Maybe we have nothing after "="
+					// or we have a double quote
+					l, _ = c.Next()
+					if l.value == zQuote {
+						// Only needed when value ends with double quotes
+						// Any value starting with zQuote ends with it
+						canHaveNextKey = false
+
+						l, _ = c.Next()
+						switch l.value {
+						case zString:
+							// We have a value in double quotes
+							val = l.token
+							l, _ = c.Next()
+							if l.value != zQuote {
+								return &ParseError{"", "svcb unterminated value", l}
+							}
+						case zQuote:
+							// There's nothing in double quotes
+						default:
+							return &ParseError{"", "svcb invalid value", l}
+						}
+					}
+				}
+			}
+			key_value = makeSVCBKeyValue(SVCBStringToKey(key))
+			if key_value == nil {
+				return &ParseError{"", "svcb invalid key", l}
+			}
+			if err := key_value.read(val); err != nil {
+				return &ParseError{"", err.Error(), l}
+			}
+			xs = append(xs, key_value)
+		case zQuote:
+			return &ParseError{"", "svcb key can't contain double quotes", l}
+		case zBlank:
+			canHaveNextKey = true
+		default:
+			return &ParseError{"", "bad svcb Values", l}
+		}
+		l, _ = c.Next()
+	}
+	rr.Value = xs
+	if rr.Priority == 0 && len(xs) > 0 {
+		return &ParseError{"", "svcb aliasform can't have values", l}
+	}
+	return nil
+}
+
+// makeSVCBKeyValue returns an SVCBKeyValue struct with the key
+// or nil for reserved keys.
+func makeSVCBKeyValue(key uint16) SVCBKeyValue {
```

Comment:
We have another option: to allow private types between `SVCB_PRIVATE_LOWER` and `SVCB_PRIVATE_UPPER` and consider the rest invalid? So that this library can be used an  as a dns proxy without needing to adjust for private types. Otherwise, do I completely remove SVCBLocal or move them to somewhere else?